### ISSUE TITLE
Enable entrystream to encode and decode MarkedSource facts [phab:D1721]

### ISSUE THREAD

**Issue by kythe** [google/kythe#2152]
_Created by **djrenren** at 2017-06-16 22:22:14_: 

Allow entrystream to handle JSON representations of MarkedSource facts.

**Comment by kythe** [comment]
_**schroederc** commented inline (context lost during migration) at 2017-06-19 16:47:09_:

Use `proto.Marshal`

**Comment by kythe** [comment]
_**schroederc** commented inline (context lost during migration) at 2017-06-19 16:47:09_:

Third-party import group should be below proto import group.

**Comment by kythe** [comment]
_**schroederc** commented inline (context lost during migration) at 2017-06-19 16:47:09_:

This flag could have a more general effect than just decoding `/kythe/edge/code` facts since that check is in the library.  The flag name should probably be more general as well.  Something like `--rich_json`?

**Comment by kythe** [comment]
_**schroederc** commented inline (context lost during migration) at 2017-06-19 16:47:09_:

Use `proto.Equal`

**Comment by kythe** [comment]
_**schroederc** required changes at 2017-06-19 16:47:09_

**Comment by kythe** [comment]
_**schroederc** commented inline (context lost during migration) at 2017-06-19 16:47:10_:

The fact name is actually `/kythe/edge/code`.  Please use the https://godoc.org/kythe.io/kythe/go/util/schema/facts library.

**Comment by kythe** [comment]
_**fromberger** commented inline (context lost during migration) at 2017-06-19 16:49:38_:

A completely cosmetic nit: Can we not use "rich" to describe this? The term is quite vague—if we're going to be general how about `decode_fact_values` or if specific `json_code` ? (Those names aren't crucial, but something more specific to the task would be nice)

**Comment by kythe** [comment]
_**djrenren** wrote at 2017-06-19 19:07:31_:

  - Rename "rich" to "structured" + test adjustments

**Comment by kythe** [comment]
_**djrenren** wrote at 2017-06-19 19:09:08_:

  - Rename "rich" to "structured" + test adjustments

**Comment by kythe** [comment]
_**fromberger** commented inline (context lost during migration) at 2017-06-19 19:11:35_:

Does this type need exported?

Also, nit: `…to handle structured fact values…`

**Comment by kythe** [comment]
_**fromberger** commented inline (context lost during migration) at 2017-06-19 19:11:35_:

Debugging?

**Comment by kythe** [comment]
_**djrenren** wrote at 2017-06-19 20:07:40_:

  - Rename "rich" to "structured" + test adjustments

Cleaned up some typos

**Comment by kythe** [comment]
_**djrenren** commented inline (context lost during migration) at 2017-06-19 20:10:44_:

Just goes to show where find and replace gets ya.

**Comment by kythe** [comment]
_**djrenren** commented inline (context lost during migration) at 2017-06-19 20:10:44_:

oops. yeah.

**Comment by kythe** [comment]
_**djrenren** commented inline (context lost during migration) at 2017-06-19 20:10:44_:

I want to leave it exported since we rely on the type externally. see the writeJSON case in entryreader

**Comment by kythe** [comment]
_**djrenren** wrote at 2017-06-19 20:10:55_:

Missed a debug line

**Comment by kythe** [comment]
_**djrenren** commented inline (context lost during migration) at 2017-06-19 20:11:34_:

Just goes to show where find and replace gets ya.

**Comment by kythe** [comment]
_**fromberger** commented inline (context lost during migration) at 2017-06-20 04:56:00_:

Probably

   var marshaler = jsonpb.Marshaler{OrigName: true}

since I think (?) everywhere else we don't use pascalCase as it typically renders.

**Comment by kythe** [comment]
_**fromberger** commented inline (context lost during migration) at 2017-06-20 04:56:00_:

Don't ignore the error this can return.

```
pbms, err := proto.Marshal(ms)
if err != nil {
  t.Fatalf("Error marshaling fact value: %v")
}
```

**Comment by kythe** [comment]
_**fromberger** commented inline (context lost during migration) at 2017-06-20 04:56:00_:

Please add context to these, since there are several steps. E.g.,

    t.Fatalf("Marshaling test entry: %v", err)

or words to that effect.

**Comment by kythe** [comment]
_**djrenren** wrote at 2017-06-20 22:11:16_:

  - Add more robust error messages and OrigName json encoding

**Comment by kythe** [comment]
_**fromberger** accepted this revision at 2017-06-20 22:49:23_

**Comment by kythe** [comment]
_**schroederc** accepted this revision at 2017-06-21 16:37:45_

**Comment by kythe** [comment]
_**schroederc** closed via commit 7e73fcc10fb5d98b5e7b0128ccd5546cf45c1129 at 2017-06-22 20:48:46_
